### PR TITLE
Add CI

### LINKED ISSUE #557: Feature: Add tests (CI)
**Problem [optional]:** Sometimes we break things, and we don't notice til it's live.

**Request:** Add basic tests

**Additional context**:

### CI:

We need to get something basic in place.  We can use GitHub Actions.

Then we can create separate issues.

For example:

Code:
- [ ] Make sure there is no 404 where we totally fail to serve the homepage (like https://github.com/machinetranslate/machinetranslate.org/pull/550).

Data (i.e. for auto-generated pages):
- [ ] Make sure that there are no pages with a conflicting path (e.g. a TMS and an API with the same slug)

Content:
- [ ] Make sure the served paths are only one level deep (e.g. /swahili not /language/swahili).
- [ ] Make sure links to pages are not broken.
- [ ] Make terms are linked to their pages where possible.
- [ ] Enforce the style guide.
   - [ ] No abbreviations
   - [ ] No US-specific spellings
   - [ ] Newline after each sentence, 2 after space.

### CD:

We basically already have this, thanks to GitHub Pages.

There is one exception, but maybe it's fine.
- [] Run generate.py automatically.
If we decide to do that, we should create a separate issue.

### PR DESCRIPTION
**Pull request by tovmasharrison**
# Description
I have added a CI and the below tests:

- Make sure there is no 404 where we totally fail to serve the homepage 
- Make sure that there are no pages with a conflicting path (e.g. a TMS and an API with the same slug)
- Make sure the served paths are only one level deep (e.g. /swahili not /language/swahili).
- Make sure links to pages are not broken (e.g. missing .md, relative path...)
- Make terms are linked to their pages where possible.

Fixes #557 partially, as tests to enforce the style guide still need to be added. 

### Checklist:

- [x] I have read the [contributing guidelines](/CONTRIBUTING).
- [x] I have followed the [style guide](http://machinetranslate.org/style).

### CODE REVIEW

**Comment by tovmasharrison** [comment]
A slight issue occurred when checking broken links. I'm working on it.

File: .github/tests/find_missing_links.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        f.seek(0)
+        next(f)
+        is_frontmatter = True
+        in_html_block = False
+
+        content = f.read()
+        for line in content.splitlines():
+
+            # Return if autogenerated
+            if "autogenerated: true" in line:
```
Comment (bittlingmayer) [inline .github/tests/find_missing_links.py:36]: stricter without being too brittle:
```
if line.strip() == "autogenerated: true":
```

File: .github/tests/find_missing_links.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                return
+
+            # Skip if frontmatter
+            if is_frontmatter:
+                if line.startswith('---'):
+                    is_frontmatter = False
+                continue
+            
+            # Remove HTML
```
Comment (bittlingmayer) [inline .github/tests/find_missing_links.py:45]:
```
# Skip HTML
```

File: .github/tests/find_missing_links.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            if '<' in line and '>' in line:
+                in_html_block = True
+
+            if in_html_block:
+                continue
+
+            _line = line.lower()
+
+            # Remove existing links/Liquid tags
```
Comment (bittlingmayer) [inline .github/tests/find_missing_links.py:54]:
```
# Remove existing Markdown links and Liquid tags
```

File: .github/tests/find_missing_links.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                # If we already have this suggestion, skip.
+                if suggestion_key in suggestions:
+                    continue
+                suggestions.add(suggestion_key)
+
+                print(f'"{article_name}" in {file_path.split("/")[-1]}')
+                print(f'\t at the line: {line}')
+                print(f'\t can be linked: [{article_name}](/{article_name.replace(" ", "-")}) \n')
+                print(30 * '-')
+                input("Press ENTER for more...") # Input is used to view each
```
Comment (bittlingmayer) [inline .github/tests/find_missing_links.py:76]: Won't this get stuck during CI?  Maybe just print them all?

File: .github/tests/test_check_urls.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import concurrent.futures
+import requests
+from bs4 import BeautifulSoup
+import pytest
+
```
Comment (bittlingmayer) [inline .github/tests/test_check_urls.py:5]: Add some comments in this file

File: .github/tests/test_requirements.txt [hunk 6]
```diff
@@ -0,0 +1,55 @@
+aiohttp==3.8.3
```
Comment (bittlingmayer) [inline .github/tests/test_requirements.txt:1]: Call it requirements.txt, more standard.

Ideally we only include those that we actually import directly.

File: .github/tests/test_slugs.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import requests
+from bs4 import BeautifulSoup
+
+
```
Comment (bittlingmayer) [inline .github/tests/test_slugs.py:4]: Add a comment here

```Test that slugs are unique

File: _config.yml [hunk 8]
```diff
@@ -234,8 +234,12 @@ social:
 
 lsi: false
 
+# Enable incremental builds by default
 incremental: true
 
+# Add GitHub Metadata
+github: [metadata]
```
Comment (bittlingmayer) [inline _config.yml:241]: What's this?  Can you add a link in the comment?

File: _layouts/language.html [hunk 9]
```diff
@@ -20,7 +20,7 @@ <h2>
 
       You can add it to Machine Translate by getting the name, language family, script and typology to <a target="_blank" href="https://github.com/machinetranslate/machinetranslate.org/blob/master/_data/languages.json">languages.json</a> and submitting a pull request.
 
-      You may be able to find this information in  <a target="_blank" href="https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes#{{ page.codes }}">its Wikipedia article.</a>
+      You may be able to find this information in  <a target="_blank" href="https://en.wikipedia.org/wiki/ISO_639:{{ page.codes }}">its Wikipedia article.</a>
```
Comment (bittlingmayer) [inline _layouts/language.html:23]: Nice.

But won't `codes` be a list, not a single code?

File: contributing/style.md [hunk 10]
```diff
@@ -1,5 +1,5 @@
 ---
-nav_order: 1
+nav_order: 1``
```
Comment (bittlingmayer) [inline contributing/style.md:2]: Bug?

File: _data/quality_estimation.json [hunk 11]
```diff
@@ -1644,7 +1644,7 @@
         ]
     },
     {
-        "id": "omniscien-qe",
+        "id": "omniscien-confidence-scores",
         "name": "Omniscien Confidence Scores",
         "company": "Omniscien-Technologies",
```
Comment (bittlingmayer) [inline _data/quality_estimation.json:1649]: Can you remove the `-` please?

Comment (bittlingmayer) [reply _data/quality_estimation.json:1649]: Also 

```
"only_compatible_mt_api": "kantanmt",
```

is not correct.

```
"only_compatible_mt_api": "omniscien",
```

File: resources/reports.md [hunk 12]
```diff
@@ -8,7 +8,7 @@ seo:
 
 |     |     |
 | --- | --- |
-| [Phrase](/phrase) | [Q2 2023 Machine Translation Report](https://phrase.com/resources/machine-translation-report/) |
+| [Phrase](/phrase-mtqe) | [Q2 2023 Machine Translation Report](https://phrase.com/resources/machine-translation-report/) |
```
Comment (bittlingmayer) [inline resources/reports.md:11]: No, should be `phrase-tms`

File: .github/tests/find_missing_links.py (hunk 4 above)
Comment (tovmasharrison) [reply .github/tests/find_missing_links.py:76]: But are we going to include this in the CI?

File: _config.yml (hunk 8 above)
Comment (tovmasharrison) [reply _config.yml:248]: The CI was failing due to "No GitHub API authentication" error. 

Adding that line solved the [issue](https://github.com/github/pages-gem/issues/399).

**Comment by tovmasharrison** [comment]
@bittlingmayer All done.

**Pull request merged**